model/jpa/src/main/java/org/keycloak/models/jpa/session/PersistentUserSessionEntity.java

```diff
@@ -47,14 +47,13 @@
                 " AND sess.realmId = :realmId AND sess.userId = :userId ORDER BY sess.userSessionId"),
         @NamedQuery(name="findUserSessionsByClientId", query="SELECT sess FROM PersistentUserSessionEntity sess INNER JOIN PersistentClientSessionEntity clientSess " +
                 " ON sess.userSessionId = clientSess.userSessionId AND clientSess.clientId = :clientId WHERE sess.offline = :offline " +
-                " AND sess.userSessionId = clientSess.userSessionId AND sess.realmId = :realmId ORDER BY sess.userSessionId"),
-        @NamedQuery(name="findUserSessionsCountsByClientId", query="SELECT clientSess.clientId, count(clientSess) " +
-                " FROM PersistentUserSessionEntity sess INNER JOIN PersistentClientSessionEntity clientSess " +
-                " ON sess.userSessionId = clientSess.userSessionId " +
-                // find all available offline user-session for all clients in a realm
-                " WHERE sess.offline = :offline " +
-                " AND sess.userSessionId = clientSess.userSessionId AND sess.realmId = :realmId " +
-                " GROUP BY clientSess.clientId")
+                " AND sess.realmId = :realmId ORDER BY sess.userSessionId"),
+        @NamedQuery(name="findUserSessionsByExternalClientId", query="SELECT sess FROM PersistentUserSessionEntity sess INNER JOIN PersistentClientSessionEntity clientSess " +
+                " ON sess.userSessionId = clientSess.userSessionId AND clientSess.clientStorageProvider = :clientStorageProvider AND clientSess.externalClientId = :externalClientId WHERE sess.offline = :offline " +
+                " AND sess.realmId = :realmId ORDER BY sess.userSessionId"),
+        @NamedQuery(name="findClientSessionsClientIds", query="SELECT clientSess.clientId, clientSess.externalClientId, clientSess.clientStorageProvider" +
+                " FROM PersistentClientSessionEntity clientSess INNER JOIN PersistentUserSessionEntity sess ON clientSess.userSessionId = sess.userSessionId " +
+                " WHERE sess.offline = :offline AND sess.realmId = :realmId")
```

Comment:
One more idea to this, I am not sure this could work this way, but I believe it is worth trying since it could have better performance.
```suggestion
        @NamedQuery(name="findClientSessionsClientIds", query="SELECT clientSess.clientId, clientSess.externalClientId, clientSess.clientStorageProvider, count(clientSess)" +
                " FROM PersistentClientSessionEntity clientSess INNER JOIN PersistentUserSessionEntity sess ON clientSess.userSessionId = sess.userSessionId " +
                " WHERE sess.offline = :offline AND sess.realmId = :realmId" +
                "GROUP BY clientSess.clientId, clientSess.externalClientId, clientSess.clientStorageProvider")
```

This way we could get rid of this:
```
closing(query.getResultStream())
                .map(row -> {
                    String clientId = row[0].toString();
                    if (clientId.equals(PersistentClientSessionEntity.EXTERNAL)) {
                        final String externalClientId = row[1].toString();
                        final String clientStorageProvider = row[2].toString();
                        clientId = new StorageId(clientStorageProvider, externalClientId).getId();
                    }
                    return clientId;
                })
                .sorted()
                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
```

and replace it with something like this:
```
closing(query.getResultStream())
                .collect(Collectors.toMap(row -> {
                    String clientId = row[0].toString();
                    if (clientId.equals(PersistentClientSessionEntity.EXTERNAL)) {
                        final String externalClientId = row[1].toString();
                        final String clientStorageProvider = row[2].toString();
                        clientId = new StorageId(clientStorageProvider, externalClientId).getId();
                    }
                    return clientId;
                }, row -> (Long) row[3])); // Not sure about the exact type of row[3] so I am not sure this will work
```

WDYT?

Reply:
I think it might work. I'll double check it with some tests. But what about ordering? Do we care about order of the entries in the resulting map by final `clientId`? If yes, we probably need ordering in Java.

Reply:
Hmm, this is a good question, I completely missed it. The previous method used HashMap for returning the result and HashMap does not preserve order so I believe the order is not needed. Also, the current order would be a little bit different as we are using `StorageId` that is producing id in the form `f:......:.......` which wasn't there before. Maybe we could get another opinion on this, whether some order is needed @hmlnarik? WDYT?

In case some order is needed, I believe it would be possible to use something like that: `ORDER BY clientSess.clientId, clientSess.clientStorageProvider`.

Reply:
Yeah, we probably don't have to care about ordering as the return type is HashMap. The method is called only from `AdminRealmResource` where the map is again transformed and return type is `Stream<Map<String, String>>`. 
I also extended one test in `ClientStorageTest` to test the query more.